saleor/discount/utils.py

```diff
@@ -1,30 +1,139 @@
+from django.conf import settings
 from django.db.models import F
+from django.utils.encoding import smart_text
+from django.utils.translation import pgettext
+from prices import FixedDiscount, Price
 
+
+from ..cart.utils import (
+    get_category_variants_and_prices, get_product_variants_and_prices)
 from .models import NotApplicable
+from . import VoucherApplyToProduct, VoucherType
 
 
 def increase_voucher_usage(voucher):
+    """Increases voucher uses by 1."""
     voucher.used = F('used') + 1
     voucher.save(update_fields=['used'])
 
 
 def decrease_voucher_usage(voucher):
+    """Decreases voucher uses by 1."""
     voucher.used = F('used') - 1
     voucher.save(update_fields=['used'])
 
 
-def get_product_discounts(product, discounts, **kwargs):
+def is_category_on_sale(category, sale):
+    """Checks if category is descendant of one of categories on sale."""
+    discounted_categories = set(sale.categories.all())
+    return any([
+        category.is_descendant_of(c, include_self=True)
+        for c in discounted_categories])
+
+
+def get_product_discount_on_sale(sale, product):
+    """Returns discount value if product is on sale or raises NotApplicable
+    elsewhere."""
+    discounted_products = {p.pk for p in sale.products.all()}
+    is_product_on_sale = (
+        product.pk in discounted_products or
+        is_category_on_sale(product.category, sale))
+    if is_product_on_sale:
+        return sale.get_discount()
+    raise NotApplicable(
+        pgettext(
+            'Voucher not applicable',
+            'Discount not applicable for this product'))
+
+
+def get_product_discounts(product, discounts):
+    """Returns discount values for all discounts applicable to a product."""
     for discount in discounts:
         try:
-            yield discount.modifier_for_product(product, **kwargs)
+            yield get_product_discount_on_sale(discount, product)
         except NotApplicable:
             pass
 
 
-def calculate_discounted_price(product, price, discounts, **kwargs):
+def calculate_discounted_price(product, price, discounts):
+    """Returns minimum product's price of all prices with discounts applied."""
     if discounts:
         discounts = list(
-            get_product_discounts(product, discounts, **kwargs))
+            get_product_discounts(product, discounts))
         if discounts:
             price = min(price | discount for discount in discounts)
     return price
+
+
+def _get_value_voucher_discount_for_checkout(voucher, checkout):
+    """Calculates discount value for a voucher of value type."""
```

Comment:
Could you validate docstrings format with `pydocstyle`?

Reply:
Sure, I'll fix it.